tasks/korolev_k_string_word_count/mpi/src/ops_mpi.cpp

```diff
@@ -0,0 +1,118 @@
+#include "korolev_k_string_word_count/mpi/include/ops_mpi.hpp"
+
+#include <mpi.h>
+
+#include <algorithm>
+#include <cctype>
+#include <cstddef>
+#include <cstdint>
+#include <string>
+
+#include "korolev_k_string_word_count/common/include/common.hpp"
+
+namespace korolev_k_string_word_count {
+
+namespace {
+
+int CountWordsChunk(const std::string &s, std::size_t begin, std::size_t end, bool prev_is_space) {
+  if (begin >= end) {
+    return 0;
+  }
+  int count = 0;
+  bool in_word = false;
+
+  auto first = static_cast<unsigned char>(s[begin]);
+  bool first_is_space = std::isspace(first) != 0;
+  if (!first_is_space && prev_is_space) {
+    ++count;
+  }
+  in_word = !first_is_space;
+
+  for (std::size_t i = begin + 1; i < end; ++i) {
+    auto c = static_cast<unsigned char>(s[i]);
+    bool is_space = std::isspace(c) != 0;
+    if (!is_space) {
+      if (!in_word) {
+        ++count;
+        in_word = true;
+      }
+    } else {
+      in_word = false;
+    }
+  }
+  return count;
+}
+
+}  // namespace
+
+KorolevKStringWordCountMPI::KorolevKStringWordCountMPI(const InType &in) {
+  SetTypeOfTask(GetStaticTypeOfTask());
+  GetInput() = in;
+  GetOutput() = 0;
+}
+
+bool KorolevKStringWordCountMPI::ValidationImpl() {
+  // Any string is valid.
+  return GetOutput() == 0;
+}
+
+bool KorolevKStringWordCountMPI::PreProcessingImpl() {
+  GetOutput() = 0;
+  return true;
+}
+
+bool KorolevKStringWordCountMPI::RunImpl() {
+  int rank = 0;
+  int size = 1;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+  std::string s;
+  uint64_t n = 0;
+
+  if (rank == 0) {
+    s = GetInput();
+    n = static_cast<uint64_t>(s.size());
+  }
+
+  MPI_Bcast(&n, 1, MPI_UINT64_T, 0, MPI_COMM_WORLD);
+
+  if (n == 0) {
+    GetOutput() = 0;
+    return true;
+  }
+
+  if (rank != 0) {
+    s.resize(static_cast<std::size_t>(n));
+  }
+  MPI_Bcast(s.data(), static_cast<int>(n), MPI_CHAR, 0, MPI_COMM_WORLD);
```

Comment:
Consider scattering the data instead of duplicating the whole string across the different ranks